### PR TITLE
fix(android/ios): amend Tab tintColor and activeTintColor

### PR DESCRIPTION
**Pull request by garymathews**
**Android**
- Amend both `Ti.UI.Tab` properties `titleColor` and `activeTitleColor`

**Android/iOS**
- Implement `tintColor` and `activeTintColor`

##### TEST CASE
```JS
const tabGroup = Ti.UI.createTabGroup({
	// style: Ti.UI.Android.TABS_STYLE_BOTTOM_NAVIGATION
});

const tabWin_home = Ti.UI.createWindow({
   	title: 'Home',
   	backgroundColor: 'white'
});
const tab_home = Ti.UI.createTab({
   	icon: 'home.png',
   	title: 'Home',
   	titleColor: 'red',
   	activeTitleColor: 'blue',
   	tintColor: 'red',
   	activeTintColor: 'blue',
   	window: tabWin_home
});

const tabWin_chat = Ti.UI.createWindow({
	title: 'Chat',
	backgroundColor: 'white'
});
const tab_chat = Ti.UI.createTab({
	icon: 'chat.png',
	title: 'Chat',
	titleColor: 'red',
	activeTitleColor: 'blue',
	tintColor: 'red',
	activeTintColor: 'blue',
	window: tabWin_chat
});

tabGroup.addTab(tab_home);
tabGroup.addTab(tab_chat);
tabGroup.open();
```
- Tab titles and text should show the appropriate colors

[TIMOB-27830](https://jira.appcelerator.org/browse/TIMOB-27830)
[TIMOB-27831](https://jira.appcelerator.org/browse/TIMOB-27831)
[TIMOB-27847](https://jira.appcelerator.org/browse/TIMOB-27847)
[TIMOB-27911](https://jira.appcelerator.org/browse/TIMOB-27911)

### CODE REVIEW

File: apidoc/Titanium/UI/Tab.yml [hunk 1]
```diff
@@ -255,6 +255,18 @@ properties:
     osver: {ios: {min: "7.0"}}
     platforms: [iphone,ipad]
 
+  - name: tintColor
+    summary: Defines the color of the tab icon.
+    type: String
+    since: {iphone: "3.1.3", ipad: "3.1.3", android: "9.1.0"}
+    platforms: [iphone, ipad, android]
+
+  - name: activeTintColor
```
Comment (hansemannn) [inline apidoc/Titanium/UI/Tab.yml:264]: This is available on iOS already. I think `tintColor` is the active tint and there is an additional one for the unselected items.

Comment (hansemannn) [reply apidoc/Titanium/UI/Tab.yml:264]:
```
'TabGroup': {
	activeTabIconTint: 'red',
	tabsTintColor: 'blue',
	unselectedItemTintColor: 'green'
}
```

Comment (sgtcoolguy) [reply apidoc/Titanium/UI/Tab.yml:264]: Looks like we have a parity issue then. We should select the property most in keeping with our typical property naming/defaults as the proper one, deprecate the other(s) and ensure parity cross-platform.

The issue here seems to be that the relevant properties do already exist on the TabGroup level, but the new implementation is at the Tab level. I can actually see a use case where you'd prefer to set some tab group default for colors but say have individual tabs use different colors when they're active, so I do think they both have merit.

Comment (jquick-axway) [reply apidoc/Titanium/UI/Tab.yml:264]: We already have a ticket written for `activeTabIconTint` here...
https://jira.appcelerator.org/browse/TIMOB-26734

If I remember right, Android didn't have a native API that supported this for top tabs at the time. It was only possible via style/theme XML settings. But there may be new APIs we can leverage now that we've migrated to AndroidX.

Comment (garymathews) [reply apidoc/Titanium/UI/Tab.yml:264]: It's kind of a mess:

###### TabGroup
- `activeTabIconTint`
- `tabsTintColor` (apparently applies to title and icons)
- `tintColor`
- `unselectedItemTintColor` (doesn't specify if this applies to title and/or icons)

###### Tab
- `tintColor`
- `activeTitleColor`
- `titleColor`

I added `tintColor` and `activeTintColor` to `Tab` as this is where the tab title colors are also defined. But there are a number of properties that can cause confusion and conflict. There's also a number of iOS specific properties that are exposed that could have cross-platform support with Android.

Comment (garymathews) [reply apidoc/Titanium/UI/Tab.yml:264]: I'm going to mark this as a WIP and will clean up parity with iOS.

Comment (hansemannn) [reply apidoc/Titanium/UI/Tab.yml:264]: Thanks Gary! Thats really worth it and a big improvement. And sorry for noising here 😄

**Review by jquick-axway — APPROVED** [review]
Java-side looks good to me.

File: apidoc/Titanium/UI/TabGroup.yml [hunk 2]
```diff
@@ -482,6 +485,24 @@ properties:
...
+  - name: activeTintColor
+    summary: The activeTintColor to apply to tabs.
+    description: |
+        This property affects the tint of selected tab icons.
+    type: String
+    since: "9.1.0"
+    default: null
+    platforms: [iphone, ipad, android]
+
+  - name: tintColor
```
Comment (vijaysingh-axway) [inline apidoc/Titanium/UI/TabGroup.yml:497]: In iOS , New introduced TabGroup properties 'activeTintColor' and 'tintColor' is not behaving same to corresponding properties 'tabsTintColor' and 'unselectedItemTintColor'. See following test case- 
`const tabGroup = Ti.UI.createTabGroup({
  // tintColor: 'green',
  // activeTintColor: 'red'

  unselectedItemTintColor: 'green',
  tabsTintColor: 'red',
});

const tabWin_home = Ti.UI.createWindow({
    title: 'Home',
    backgroundColor: 'white'
});
const tab_home = Ti.UI.createTab({
    icon: 'tab1.png',
    title: 'Home',
    window: tabWin_home
});

const tabWin_chat = Ti.UI.createWindow({
  title: 'Chat',
  backgroundColor: 'white'
});
const tab_chat = Ti.UI.createTab({
  icon: 'tab2.png',
  title: 'Chat',
  window: tabWin_chat
});

tabGroup.addTab(tab_home);
tabGroup.addTab(tab_chat);
tabGroup.open();`

Probably you have to implement these properties similar to [tabsTintColor](https://github.com/appcelerator/titanium_mobile/blob/b890f7c7743b8e38292c5d669ed502996b665a20/iphone/Classes/TiUITabGroup.m#L368) and [unselectedItemTintColor](https://github.com/appcelerator/titanium_mobile/blob/b890f7c7743b8e38292c5d669ed502996b665a20/iphone/Classes/TiUITabGroup.m#L373) .

File: iphone/Classes/TiUITabProxy.m [hunk 3]
```diff
@@ -576,31 +583,57 @@ - (void)updateTabBarItem
...
   }
+  [rootController setTitle:title];
+  UITabBarItem *ourItem = nil;
 
   systemTab = NO;
   ourItem = [[[UITabBarItem alloc] initWithTitle:title image:image selectedImage:activeImage] autorelease];
 
   TiColor *titleColor = [TiUtils colorValue:[self valueForKey:@"titleColor"]];
+  if (titleColor == nil) {
+    titleColor = [TiUtils colorValue:[tabGroup valueForKey:@"titleColor"]];
```
Comment (vijaysingh-axway) [inline iphone/Classes/TiUITabProxy.m:628]: tabGroup do not have titleColor property.

File: iphone/Classes/TiUITabProxy.m [hunk 4]
```diff
@@ -576,31 +583,57 @@ - (void)updateTabBarItem
...
+  }
   if (titleColor != nil) {
     [ourItem setTitleTextAttributes:[NSDictionary dictionaryWithObjectsAndKeys:[titleColor color], NSForegroundColorAttributeName, nil] forState:UIControlStateNormal];
   }
   TiColor *activeTitleColor = [TiUtils colorValue:[self valueForKey:@"activeTitleColor"]];
+  if (activeTitleColor == nil) {
+    activeTitleColor = [TiUtils colorValue:[tabGroup valueForKey:@"activeTitleColor"]];
```
Comment (vijaysingh-axway) [inline iphone/Classes/TiUITabProxy.m:635]: tabGroup do not have 'activeTitleColor' property. No need to have this line.

**Review by vijaysingh-axway — CHANGES_REQUESTED** [review]
@garymathews There are few issues in iOS side. Please address that.

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/tabgroup/TiUIBottomNavigationTabGroup.java [hunk 5]
```diff
@@ -168,12 +169,10 @@ public void addTabItemInController(TiViewProxy tabProxy)
...
-		if (this.proxy.hasPropertyAndNotNull(TiC.PROPERTY_SHIFT_MODE)) {
-			if (!((Boolean) proxy.getProperty(TiC.PROPERTY_SHIFT_MODE))) {
-				this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_LABELED);
-			} else {
-				this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_SELECTED);
-			}
+		if (proxy.getProperties().optBoolean(TiC.PROPERTY_SHIFT_MODE, true)) {
+			this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_AUTO);
+		} else {
+			this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_UNLABELED);
```
Comment (jquick-axway) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/tabgroup/TiUIBottomNavigationTabGroup.java:175]: When "shiftMode" is set `false`, we're supposed to show titles for all tabs. Even if the tab is not selected. That was the old behavior. Most devs are setting it to `false` with the intention of showing titles. So, we should change it back to `LABEL_VISIBILITY_SELECTED`.

That said, if all tabs are not assigned a "title", then perhaps we should ignore the "shiftMode" property and set it `LABEL_VISIBILITY_UNLABELED`? That way devs can have icon-only tabs without the empty space reserved for the title.

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/tabgroup/TiUIAbstractTabGroup.java [hunk 6]
```diff
@@ -425,6 +454,47 @@ public void updateTitle(String title)
...
 
+	public Drawable updateIconTint(TiViewProxy tabProxy, Drawable drawable, boolean selected)
+	{
+		final KrollDict tabProperties = tabProxy.getProperties();
+		final KrollDict properties = getProxy().getProperties();
+
+		if (drawable == null) {
+			final String title = tabProperties.optString(TiC.PROPERTY_TITLE, "");
+			Log.w(TAG, "Could not update tab icon tint color for '" + title + "' tab.");
+			return null;
```
Comment (jquick-axway) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/tabgroup/TiUIAbstractTabGroup.java:465]: Can you remove the "Could not update tab icon tint color..." warning log message please?
If I don't set any tab icons, then I end up seeing warning messages every time I select a tab.
Note that I think the top tab style looks fine without icons.

**Review by jquick-axway — CHANGES_REQUESTED** [review]
Requesting 2 changes:
- Remove the "Could not update tab icon tint color..." warning message.
- Setting "shiftMode" to `false` should show all tab titles.

File: iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m [hunk 7]
```diff
@@ -768,6 +768,25 @@ + (UIImage *)adjustRotation:(UIImage *)image
...
+    imageView.tintColor = tintColor;
+
+    UIGraphicsBeginImageContextWithOptions(imageView.bounds.size, NO, 0.0);
+    [imageView.layer renderInContext:UIGraphicsGetCurrentContext()];
+    UIImage *tintedImage = UIGraphicsGetImageFromCurrentImageContext();
+    UIGraphicsEndImageContext();
+
+    return tintedImage;
+  }
+  return image;
```
Comment (vijaysingh-axway) [inline iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m:787]: This statement will never get executed. Remove this line.

File: iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m [hunk 8]
```diff
@@ -768,6 +768,25 @@ + (UIImage *)adjustRotation:(UIImage *)image
   return imageCopy;
 }
 
++ (UIImage *)imageWithTint:(UIImage *)image tintColor:(UIColor *)tintColor
+{
+  if (![TiUtils isIOSVersionOrGreater:@"13.0"]) {
```
Comment (vijaysingh-axway) [inline iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m:773]: This condition is wrong. Remove negation.

File: iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m [hunk 9]
```diff
@@ -768,6 +768,25 @@ + (UIImage *)adjustRotation:(UIImage *)image
...
+    return [image imageWithTintColor:tintColor renderingMode:UIImageRenderingModeAlwaysOriginal];
```
Comment (vijaysingh-axway) [inline iphone/TitaniumKit/TitaniumKit/Sources/API/TiUtils.m:774]: This API is available from iOS sdk 13 + only. So building on Xcode < 11 will fail. This whole function should be like -
`+ (UIImage *)imageWithTint:(UIImage *)image tintColor:(UIColor *)tintColor
{
  #if __IPHONE_OS_VERSION_MAX_ALLOWED >= 130000
  if ([TiUtils isIOSVersionOrGreater:@"13.0"]) {
    return [image imageWithTintColor:tintColor renderingMode:UIImageRenderingModeAlwaysOriginal];
  } else {
#endif
    UIImage *imageNew = [image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    UIImageView *imageView = [[UIImageView alloc] initWithImage:imageNew];
    imageView.tintColor = tintColor;

    UIGraphicsBeginImageContextWithOptions(imageView.bounds.size, NO, 0.0);
    [imageView.layer renderInContext:UIGraphicsGetCurrentContext()];
    UIImage *tintedImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();

    return tintedImage;
  #if __IPHONE_OS_VERSION_MAX_ALLOWED >= 130000
  }
  #endif
}
`

**Review by vijaysingh-axway — APPROVED** [review]
iOS changes looks good.

**Comment by jquick-axway** [comment]
@garymathews, with your newest changes the bottom tab titles never appear. I've tried setting "shiftMode" to `false`, `true`, `0`, etc. None of these settings have an affect anymore.
```javascript
function createTab(title) {
	var window = Ti.UI.createWindow({ title: title });
	window.add(Ti.UI.createLabel({ text: title + " View" }));
	var tab = Ti.UI.createTab({
//		icon: "/assets/images/tab1.png",
		title: title,
		window: window,
	});
	return tab;
}

var tabGroupSettings = {
	tabs: [createTab("Tab 1"), createTab("Tab 2"), createTab("Tab 3")],
//	tabsBackgroundColor: "blue",
//	tabsBackgroundSelectedColor: "blue",
};
if (Ti.Android) {
	tabGroupSettings.shiftMode = false;
	tabGroupSettings.style = Ti.UI.Android.TABS_STYLE_BOTTOM_NAVIGATION;
}
var tabGroup = Ti.UI.createTabGroup(tabGroupSettings);
tabGroup.open();
```

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/tabgroup/TiUIBottomNavigationTabGroup.java [hunk 10]
```diff
@@ -169,10 +169,15 @@ public void addTabItemInController(TiViewProxy tabProxy)
...
-			this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_UNLABELED);
+		final int shiftMode = proxy.getProperties().optInt(TiC.PROPERTY_SHIFT_MODE, 1);
+		switch (shiftMode) {
+			case 0:
+				this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_SELECTED);
+			case 1:
+				this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_AUTO);
+			case 2:
+				// NOTE: Undocumented for now, will create new property that has parity with iOS.
+				this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_UNLABELED);
```
Comment (jquick-axway) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/tabgroup/TiUIBottomNavigationTabGroup.java:180]: You are missing `break` statements. Also, `case 0` needs to be set to `LABELED` to match the old behavior which always shows tab title. Setting it to `SELECTED` means it only shows a tab title when selected which is what `case 1` is supposed to do.
```java
switch (shiftMode) {
	case 0:
		this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_LABELED);
		break;
	case 1:
		this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_AUTO);
		break;
	case 2:
		// NOTE: Undocumented for now, will create new property that has parity with iOS.
		this.mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_UNLABELED);
		break;
}
```

**Review by jquick-axway — CHANGES_REQUESTED** [review]

**Review by jquick-axway — APPROVED** [review]
CR: Pass

**Review by ssjsamir — APPROVED** [review]
FR Passed: Tested with the test case above, able to see the tint colours on titles and icons. 

**Test Environment**
```
MacOS Catalina: 10.15.5 Beta
Xcode: 11.4
Java Version: 1.8.0_131
Android NDK: 21.1.6273396-beta2
Node.js: 10.16.3
""NPM":"5.0.0","CLI":"8.0.0""
Andoird Pixl xl 7.1.1 Emulator 
iphone 8 13.4 sim
```

**Comment by sgtcoolguy** [comment]
Manually merged to master. There was a small conflict preventing my merge/rebase, so had to do it locally.

**Comment by energy2080** [comment]
This issue is back in latest 9.1.0.v20200824195336

**Pull request closed without merging**